Patches: Adds db_nodes_autoinc patch

tomponline opened the pull request:
Recreates nodes table id column with AUTOINCREMENT.

Related to https://github.com/canonical/raft/pull/176
Related to https://github.com/lxc/lxd/pull/8521

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

tomponline commented on lxd/patches.go:
```diff
@@ -181,6 +182,51 @@ func patchesApply(d *Daemon, stage patchStage) error {
+		return err
+	}
+
+	if strings.Contains(schemaSQL, "id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL") {
```
@stgraber not sure if this is what you had in mind to only run on leader, but this should prevent it running multiple times.

stgraber replied:
That's a good protection to have.

My thought was to have call `node.ClusterAddress(d.db)` combined with `d.gateway.LeaderAddress()` and if they're not equal, then skip as we're not on the leader.

stgraber commented on lxd/patches.go:
```diff
@@ -181,6 +182,51 @@ func patchesApply(d *Daemon, stage patchStage) error {
+	}
+
+	_, err = d.State().Cluster.DB().Exec(`
+PRAGMA foreign_keys = OFF;
+PRAGMA legacy_alter_table = ON;
```
Isn't that global? If so, you definitely need to revert that at the end.

tomponline replied:
Mmm yeah if the connection is held open after patches then yes. I'll also add a check for leader so there's no chance of a race between checking the existing schema and applying the fix if missing.

tomponline commented:
@stgraber I've updated this with the changes suggested, and have modified the way that the patch works to loop waiting for the patch to be applied on the leader (if clustered), because in my testing on a cluster, I found that if there was an error applying the patch (in my case I'd made a typo in the SQL) the leadership seemed to flip flop and eventually allowed the patch to be skipped and considered "applied" on all nodes due to the fact that at some point each of the nodes run the patch when not a leader and skipped apply it.

Now I've updated the patch so that each node checks if the schema needs the patch, and if so, then only applies it if its the leader, and if not waits for 1s and then checks the schema again, and repeats. 

Now in my test I see this:

Node 1:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:19+0000 lvl=warn msg="Waiting for \"db_nodes_autoinc\" patch to be applied on leader cluster member" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=dbug msg="Skipping \"db_nodes_autoinc\" patch as \"nodes\" table id column already AUTOINCREMENT" 
t=2021-03-05T10:13:20+0000 lvl=info msg="Initializing daemon storage mounts" 
```

Node 2:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=dbug msg="Found cert" name=0
t=2021-03-05T10:13:20+0000 lvl=info msg="Initializing daemon storage mounts" 
```

Node 3:

```
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Initializing and checking storage pool \"local\"" 
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount started" driver=dir pool=local
t=2021-03-05T10:13:19+0000 lvl=dbug msg="Mount finished" driver=dir pool=local
t=2021-03-05T10:13:20+0000 lvl=info msg="Applying patch \"db_nodes_autoinc\"" 
t=2021-03-05T10:13:20+0000 lvl=warn msg="Waiting for \"db_nodes_autoinc\" patch to be applied on leader cluster member" 
t=2021-03-05T10:13:21+0000 lvl=dbug msg="Skipping \"db_nodes_autoinc\" patch as \"nodes\" table id column already AUTOINCREMENT" 
t=2021-03-05T10:13:21+0000 lvl=info msg="Initializing daemon storage mounts" 
```

So it applied on node 2, and then the other two noticed and proceeded.

tomponline commented:
jenkins: test this please
